### PR TITLE
Use new hosting model for the app template.

### LINKED ISSUE #6828: Async application initialization methods on the abpmodule class
I had closed this before: https://github.com/abpframework/abp/issues/5382
I am opening again and adding to backlog for next major versions.

Also, see https://github.com/maliming/AbpAsyncApplicationLifecycle

### LINKED ISSUE #10426: Consider using the new hosting model
It is easy to apply for who want to switch today: https://github.com/abpframework/abp/commit/777c3ddd7a69289d2434acf16b90c7db2e8dc36f

We want to wait until it is well understood. Some developers may ask "where is my Startup class?" 😄

### PR DESCRIPTION
**Pull request by maliming**
Resolve #6828
Resolve #10426

## Changes:
1. The Module class.

The `PreConfigureServices`, `ConfigureServices`, `PostConfigureServices`, `OnPreApplicationInitialization`, `OnApplicationInitialization`, `OnPostApplicationInitialization` methods of module have the `Async` version now. If you need to use an async method, please override the corresponding method. Async methods automatically call sync methods by default.

https://github.com/abpframework/abp/blob/new-hosting-model/framework/src/Volo.Abp.Core/Volo/Abp/Modularity/AbpModule.cs#L34-L109

2. Unit Test.

Add the new `AbpAsyncIntegratedTest` and `AbpAspNetCoreAsyncTestBase`.

The test class needs to implement the `IAsyncLifetime` interface, which has been implemented by `AbpAspNetCoreAsyncTestBase`.

Demo test classes:
https://github.com/abpframework/abp/blob/new-hosting-model/framework/test/Volo.Abp.AspNetCore.Mvc.Tests/Volo/Abp/AspNetCore/Mvc/AbpAspNetCoreAsyncTestBase_Tests.cs

https://github.com/abpframework/abp/blob/new-hosting-model/framework/test/Volo.Abp.Core.Tests/Volo/Abp/DynamicProxy/AbpInterceptionTestBase.cs

3. MVC
```cs
var builder = WebApplication.CreateBuilder(args);
builder.Host.AddAppSettingsSecretsJson()
    .UseAutofac();
await builder.Services.AddApplicationAsync<MyProjectNameWebModule>(options =>
{
    options.Services.ReplaceConfiguration(builder.Configuration);
});
var app = builder.Build();
await app.InitializeApplicationAsync();
await app.RunAsync();
```

4. Blazor WASM
```cs
 var builder = WebAssemblyHostBuilder.CreateDefault(args);
var application = await builder.AddApplicationAsync<MyProjectNameBlazorModule>(options =>
{
    options.UseAutofac();
});
var host = builder.Build();
await application.InitializeApplicationAsync(host.Services);
await host.RunAsync();
```

## Breaking changes:

If you use async methods in your module, please keep the same sync methods for compatibility.

eg: https://github.com/abpframework/abp/blob/new-hosting-model/modules/identityserver/src/Volo.Abp.IdentityServer.Domain/Volo/Abp/IdentityServer/AbpIdentityServerDomainModule.cs#L137-L163

Or 

```cs
public async override Task OnApplicationInitializationAsync(ApplicationInitializationContext context)
{
    var options = context.ServiceProvider.GetRequiredService<IOptions<TokenCleanupOptions>>().Value;
    if (options.IsCleanupEnabled)
    {
        await context.ServiceProvider
            .GetRequiredService<IBackgroundWorkerManager>()
            .AddAsync(
                context.ServiceProvider
                    .GetRequiredService<TokenCleanupBackgroundWorker>()
            );
    }
}

public override void OnApplicationInitialization(ApplicationInitializationContext context)
{
    AsyncHelper.RunSync(() => OnApplicationInitializationAsync(context));
}
```

### CODE REVIEW

File: framework/src/Volo.Abp.AspNetCore/Microsoft/AspNetCore/Builder/AbpApplicationBuilderExtensions.cs [hunk 1]
```diff
@@ -18,6 +19,27 @@ public static class AbpApplicationBuilderExtensions
...
+    {
+        Check.NotNull(app, nameof(app));
+
+        app.ApplicationServices.GetRequiredService<ObjectAccessor<IApplicationBuilder>>().Value = app;
+        var application = app.ApplicationServices.GetRequiredService<IAbpApplicationWithExternalServiceProvider>();
+        var applicationLifetime = app.ApplicationServices.GetRequiredService<IHostApplicationLifetime>();
+
+        applicationLifetime.ApplicationStopping.Register(() =>
+        {
+            application.ShutdownAsync();
```
Comment (hikalkan) [inline framework/src/Volo.Abp.AspNetCore/Microsoft/AspNetCore/Builder/AbpApplicationBuilderExtensions.cs]: Calling an async method without await or AsyncHelper.RunSync. Isn't that a problem?

File: framework/src/Volo.Abp.Core/Volo/Abp/AbpApplicationCreationOptions.cs [hunk 2]
```diff
@@ -19,10 +19,13 @@ public class AbpApplicationCreationOptions
     [NotNull]
     public AbpConfigurationBuilderOptions Configuration { get; }
 
+    public bool ManualConfigureServices { get; set; }
```
Comment (hikalkan) [inline framework/src/Volo.Abp.Core/Volo/Abp/AbpApplicationCreationOptions.cs]: What is that used for?

Comment (maliming) [reply framework/src/Volo.Abp.Core/Volo/Abp/AbpApplicationCreationOptions.cs]: The constructor of `AbpApplicationBase `will call the sync `ConfigureServices`, async initialization cannot be done in the constructor.

Comment (maliming) [reply framework/src/Volo.Abp.Core/Volo/Abp/AbpApplicationCreationOptions.cs]: image

File: framework/src/Volo.Abp.AspNetCore/Microsoft/AspNetCore/Builder/AbpApplicationBuilderExtensions.cs (hunk 1 above)
Comment (maliming) [reply framework/src/Volo.Abp.AspNetCore/Microsoft/AspNetCore/Builder/AbpApplicationBuilderExtensions.cs]: code from Microsoft, but I will confirm that.

https://docs.microsoft.com/en-us/dotnet/standard/threading/how-to-register-callbacks-for-cancellation-requests#example

**Comment by rqx110** [comment]
suggestion modifiers declaration order:
```diff
- public async static Task<int> Main(string[] args)
+ public static async Task<int> Main(string[] args)
```

```diff
- public async override Task OnApplicationInitializationAsync
+ public override async Task OnApplicationInitializationAsync

File: framework/src/Volo.Abp.AspNetCore/Microsoft/AspNetCore/Builder/AbpApplicationBuilderExtensions.cs (hunk 1 above)
Comment (maliming) [reply framework/src/Volo.Abp.AspNetCore/Microsoft/AspNetCore/Builder/AbpApplicationBuilderExtensions.cs]: https://github.com/abpframework/abp/pull/10928/commits/59284b463ad9483e606618bd53acfaafc7829bc0

**Review by hikalkan — APPROVED** [review]

**Pull request merged**